argcomplete/bash_completion.d/python-argcomplete.sh

```diff
@@ -34,21 +47,21 @@ _python_argcomplete_global() {
             else
                 return
             fi
-        elif [[ -f "${COMP_WORDS[1]}" ]] && (head -c 1024 "${COMP_WORDS[1]}" | grep --quiet "PYTHON_ARGCOMPLETE_OK") >/dev/null 2>&1; then
+        elif [[ -f "${COMP_WORDS[1]}" ]] && __python_argcomplete_scan_head_noerr "${COMP_WORDS[1]}"; then
             local ARGCOMPLETE=2
         else
             return
         fi
-    elif which "$executable" >/dev/null 2>&1; then
-        local SCRIPT_NAME=$(which "$executable")
+    elif type -P "$executable" >/dev/null 2>&1; then
+        local SCRIPT_NAME=$(type -P "$executable")
         if (type -t pyenv && [[ "$SCRIPT_NAME" = $(pyenv root)/shims/* ]]) >/dev/null 2>&1; then
             local SCRIPT_NAME=$(pyenv which "$executable")
         fi
-        if (head -c 1024 "$SCRIPT_NAME" | grep --quiet "PYTHON_ARGCOMPLETE_OK") >/dev/null 2>&1; then
+        if __python_argcomplete_scan_head_noerr "$SCRIPT_NAME"; then
             local ARGCOMPLETE=1
-        elif [[ "$(head -n 1 "$SCRIPT_NAME")" =~ ^#!(.*)$ ]] && [[ "${BASH_REMATCH[1]}" =~ ^.*(python|pypy)[0-9\.]*$ ]]; then
+        elif __python_argcomplete_scan_head_noerr "$SCRIPT_NAME" "^#!(.*)$" -n && [[ "${BASH_REMATCH[1]}" =~ ^.*(python|pypy)[0-9\.]*$ ]]; then
```

Comment:
Consider using single quotes where you don't explicitly want variable expansion etc. This regular expression works within the script, but would confuse somebody trying to replicate it in an interactive shell as the ! will cause history expansion.

Reply:
Changed to single quotes.